### DISCUSSION TITLE
How to find out it is Wayland or X system?

### DISCUSSION THREAD

**Discussion by Archie-2021** [twpayne/chezmoi#3560] [Q&A]
Is there a way to get which one is running on the system?

**Comment by halostatue** [comment]
This is not really a chezmoi question, as this is well-answered on [Stack Overflow](https://unix.stackexchange.com/a/371164). However, there are some complicating factors to *how* you would run such a complex command-line from within a chezmoi template.

NOTE: I don't use Linux as my daily driver, so I don't know whether the commands here will actually work, as I have nothing to test them on.

Here's the command-line we should use from the SO post in question:

```sh
loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type
```

First, you will need to use the `output` template function, which requires each command be quoted, but does not know how to run sub-shells or piped commands. This could be written as:

```
{{ output "bash" "-c" "'loginctl show-session $(loginctl | grep $(whoami) | awk \\'{ print $1 }\\') -p Type'" }}
```

This will *probably* work, but the levels of quoting required are particularly difficult to read, so the following (more complex) template should also work:

```
{{- $whoami := output "whoami" | trim -}}{{/* $(whoami) */}}
{{- $sessions := output "loginctl" | jq "split(\"\\n\")" | first - }}{{/* loginctl | */}}
{{- $mySessionPattern := printf "map(select(match(\"%s\"))).[0]" $whoami -}}
{{- $mySession = jq $mySessionPattern $sessions | splitList " " | first -}}{{/* grep $whoami | awk print $1 */}}
{{- output "loginctl" "show-session" $mySession "-p" "Type" -}}
```

There are other ways to break this down using `printf` and multiple executions of `bash -c "…"`.

Your other option is to make a shell script that runs `loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type
` and then just do `output "my-current-session-type"` in your template to get that value.

**Reply by Archie-2021** [reply]
Thanks, could not find running bash commands in documentation so I though I should ask for an internal command.

This opens up a lot of options and more uses out of this amazing app.

**Comment by Archie-2021** [comment]
> This is not really a chezmoi question, as this is well-answered on [Stack Overflow](https://unix.stackexchange.com/a/371164). However, there are some complicating factors to _how_ you would run such a complex command-line from within a chezmoi template.
> 
> NOTE: I don't use Linux as my daily driver, so I don't know whether the commands here will actually work, as I have nothing to test them on.
> 
> Here's the command-line we should use from the SO post in question:
> 
> ```shell
> loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type
> ```
> 
> First, you will need to use the `output` template function, which requires each command be quoted, but does not know how to run sub-shells or piped commands. This could be written as:
> 
> ```
> {{ output "bash" "-c" "'loginctl show-session $(loginctl | grep $(whoami) | awk \\'{ print $1 }\\') -p Type'" }}
> ```
> 
> This will _probably_ work, but the levels of quoting required are particularly difficult to read, so the following (more complex) template should also work:
> 
> ```
> {{- $whoami := output "whoami" | trim -}}{{/* $(whoami) */}}
> {{- $sessions := output "loginctl" | jq "split(\"\\n\")" | first - }}{{/* loginctl | */}}
> {{- $mySessionPattern := printf "map(select(match(\"%s\"))).[0]" $whoami -}}
> {{- $mySession = jq $mySessionPattern $sessions | splitList " " | first -}}{{/* grep $whoami | awk print $1 */}}
> {{- output "loginctl" "show-session" $mySession "-p" "Type" -}}
> ```
> 
> There are other ways to break this down using `printf` and multiple executions of `bash -c "…"`.
> 
> Your other option is to make a shell script that runs `loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type ` and then just do `output "my-current-session-type"` in your template to get that value.

Hey, I finally found sometime to mess with this config. I put it in a script and it work when I use `output` command. But I need to put it in an if statement and mix it with with `if` and `and` commands. I tried many ways including of course using round brackets but it didn't work. Can you try to see how I can check the returned value?

Thanks

**Reply by twpayne** [reply]
> Can you try to see how I can check the returned value?

You need to show what you tried.

**Reply by Archie-2021** [reply]
> > Can you try to see how I can check the returned value?
> 
> You need to show what you tried.

I tried something like this:

```
{{ if ( eq ( ( output "my-script.sh") "Type=x11" ) ) }}
```


I used parentheses to contain it, but this if statement value is false every time, and if I remove it it just try to interpret "Type=x11" part as shell command s and errors.

**Reply by halostatue** [reply]
I have a couple of suggestions. For investigation, remember `chezmoi execute-template`. In this case, `chezmoi execute-template '{{ output "my-script.sh" }}'`.

My *immediate* thought is that this is failing because `chezmoi` doesn't know where to find `my-script.sh`. You *probably* want to put this in `.chezmoi.workingTree` (that’s your chezmoi git directory, and is not the same as `.chezmoi.sourceDir`, which may be affected by the presence of `.chezmoiroot` in the working tree) and run this as `output (joinPath .chezmoi.workingTree "my-script.sh")`. I would also suggest that you give the script a better name. It's also likely that it would be better to put the comparison *in* your script. Assuming that it’s more or less what was done before:

```bash
#! /usr/bin/env bash

set -euo pipefail

session="$(loginctl | grep "${USER:-$(whoami)} | awk '{ print $1 }')"
session_type="$(loginctl show-session "${session}" -p Type)"

[[ "${1:-}" == "${session_type}" ]] && echo true
```

This would now be called as ~`if output (joinPath .chezmoi.workingTree "login-session-type-is" "Type=x11") …`~ `if output (joinPath .chezmoi.workingTree "login-session-type-is") "Type=x11"` assuming that you name the script `login-session-type-is` and put it in `.chezmoi.workingTree`. In Go, `""` is a false, so because the script doesn't output anything if `$1` doesn't match the discovered `$session_type`, it would fall to the `else` condition.

I reiterate: take my suggestions here with a huge grain of salt. I don't have a Linux desktop machine at all, so I can't say that what I have written will actually work in your case. But the *logic* involved I can attest to. As Knuth said, "Beware of bugs in the above code; I have only proved it correct, not tried it".

**Comment by Archie-2021** [comment]
> I have a couple of suggestions. For investigation, remember `chezmoi execute-template`. In this case, `chezmoi execute-template '{{ output "my-script.sh" }}'`.
> 
> My _immediate_ thought is that this is failing because `chezmoi` doesn't know where to find `my-script.sh`. You _probably_ want to put this in `.chezmoi.workingTree` (that’s your chezmoi git directory, and is not the same as `.chezmoi.sourceDir`, which may be affected by the presence of `.chezmoiroot` in the working tree) and run this as `output (joinPath .chezmoi.workingTree "my-script.sh")`. I would also suggest that you give the script a better name. It's also likely that it would be better to put the comparison _in_ your script. Assuming that it’s more or less what was done before:
> 
> ```shell
> #! /usr/bin/env bash
> 
> set -euo pipefail
> 
> session="$(loginctl | grep "${USER:-$(whoami)} | awk '{ print $1 }')"
> session_type="$(loginctl show-session "${session}" -p Type)"
> 
> [[ "${1:-}" == "${session_type}" ]] && echo true
> ```
> 
> This would now be called as `if output (joinPath .chezmoi.workingTree "login-session-type-is" "Type=x11") …` assuming that you name the script `login-session-type-is` and put it in `.chezmoi.workingTree`. In Go, `""` is a false, so because the script doesn't output anything if `$1` doesn't match the discovered `$session_type`, it would fall to the `else` condition.
> 
> I reiterate: take my suggestions here with a huge grain of salt. I don't have a Linux desktop machine at all, so I can't say that what I have written will actually work in your case. But the _logic_ involved I can attest to. As Knuth said, "Beware of bugs in the above code; I have only proved it correct, not tried it".

Hey, and thanks for your help again

I don't think the path is the issue. I have tried `{{ output "myscript" }}` with `chezmoi execute-template` command you mentioned before and it was printing the stdout perfectly. Altho it is a good idea to do the all stuff in bash including testing with if I don't think it is the problem. I think it can not pars the commands in order. I tried what you have suggested (put myscript in `~/.local/share/chezmoi` but it is giving me this error ( It tried to Joinpath the argumnet for script to path too)

```
chezmoi: template: stdin:1:6: executing "stdin" at <output (joinPath .chezmoi.workingTree "test-mysc000" "Type=x11")>: error calling output: ~/.local/share/chezmoi/test-mysc000/Type=x11: fork/exec ~/.local/share/chezmoi/test-mysc000/Type=x11: not a directory
```

It is the template I tested on it:

```
{{ if output (joinPath .chezmoi.workingTree "test-mysc000" "Type=x11") }}
Printing True
{{ else }}
Printing False
{{ end }}
```

**Reply by halostatue** [reply]
Ah. There’s a typo in what I wrote (as I said, I hadn’t run any of it):

```
{{ if output (joinPath .chezmoi.workingTree "test-mysc000") "Type=x11" }}
Printing True
{{ else }}
Printing False
{{ end }}
```

**Comment by Archie-2021** [comment]
> Ah. There’s a typo in what I wrote (as I said, I hadn’t run any of it):
> 
> ```
> {{ if output (joinPath .chezmoi.workingTree "test-mysc000") "Type=x11" }}
> Printing True
> {{ else }}
> Printing False
> {{ end }}
> ```

Excellent, you were right. The issue was finding path and now it is works like a charm.

Thank you very much for your help

**Comment by trinitronx** [comment]
Another way is to check for the existence of `wayland-scanner` or `XDG_SESSION_TYPE` equal to "`wayland`" on Linux.  This works because on most distros, `wayland-scanner` is installed alongside `libwayland` or at least `$XDG_SESSION_TYPE` is set.  For example:

In `.chezmoi.yaml.tmpl`, set a data variable to hold `XDG_SESSION_TYPE`:

```yaml
{{- $xdg_session_type := "" -}}
{{- if hasKey . "xdg_session_type" -}}
{{-   $xdg_session_type = .xdg_session_type -}}
{{- else if (env "XDG_SESSION_TYPE") -}}
{{-   $xdg_session_type = env "XDG_SESSION_TYPE" -}}
{{- end -}}

# ... other stuff ...

data:
  xdg_session_type: "{{ $xdg_session_type }}"
```

Then, in a template:

```go
{{ if (eq .chezmoi.os "linux") | and ((lookPath "wayland-scanner") | or (eq .xdg_session_type "wayland")) -}}
# ... wayland specific stuff ...
{{ else -}}
# ... other non-Wayland stuff ...
{{ end -}}
```

Note that only `$XDG_SESSION_TYPE` is tied to whether the _current_ session is Wayland or not.  Meanwhile. the `wayland-scanner` binary could still exist on a system with Wayland and other desktops (KDE, GNOME, Cinnamon, etc...) if the current session was something else.

For passively installing configs on a system that _might_ run Wayland sessions, using the above check is usually fine in practice.  For things that depend on the _current_ session being Wayland actively, (_warning: non-hermetic_ / tied to _current_ session at runtime) the `$XDG_SESSION_TYPE` check alone would work.
